### PR TITLE
use static_library's pic calculation instead of recalculating

### PR DESCRIPTION
**Pull request by dcbaker**
`static_library` has already calculated pic for us, so let's use that instead of re-calculating.

This was based on some issues I noticed in my build.py typed_kwargs series

### CODE REVIEW

File: mesonbuild/interpreter/interpreter.py [hunk 1]
```diff
@@ -3059,19 +3051,19 @@ def build_both_libraries(self, node, args, kwargs):
...
-            # Replace sources with objects from the shared library to avoid
-            # building them twice. We post-process the static library instead of
+            # Replace sources with objects from the static library to avoid
+            # building them twice. We post-process the shared library instead of
             # removing sources from args because sources could also come from
             # any InternalDependency, see BuildTarget.add_deps().
-            static_lib.objects.append(build.ExtractedObjects(shared_lib, shared_lib.sources, shared_lib.generated, []))
-            static_lib.sources = []
-            static_lib.generated = []
+            shared_lib.objects = [static_lib.extract_all_objects()]
```
Comment (xclaesse) [inline mesonbuild/interpreter/interpreter.py:3061]: It was intentional to not use extract_all_objects() here, see https://github.com/mesonbuild/meson/pull/10843.

Any reason you invert shared/static here?

Comment (dcbaker) [reply mesonbuild/interpreter/interpreter.py:3061]: There is no rational for why we don't use extract_all_objects in that which I see in your dfescription, just that you empty the sources because otherwise we can end up with sources from dependencies copied in.

I want to do this because I'm trying to get to the point that we don't create two targets with all of the same inputs, and then trying to mutate them afterwards, but instead create a dependent target that is not mutated.

Yes, there are two reasons
1) Conceptually you can --link-whole a static archive into a shared library, but you can't do that in reverse. That has the advantage of being able to be written like this:
```meson
shared_lib = self.build_target(('name', []), {'link_whole': static_lib})
```
Which I didn't do here because trying to figure out what arguments to pass without proper type annotations is just painful, so mostly laziness.
2) since we're taking the pic value from the static_library, it makes more sens to use its objects as well

Comment (xclaesse) [reply mesonbuild/interpreter/interpreter.py:3061]: https://github.com/mesonbuild/meson/pull/10843#issuecomment-1256449270

> CI failure is interesting, I'm impressed we have a test for that: With both_libraries(), link_with to an uninstalled static library is converted by the static library by adding to self.objects, but the shared lib does not do that. So when we do static_lib.objects = [shared_lib.extract_all_objects()] we lose some objets that we actually need.

Hmm, but since you now extract from the static lib, I guess that works... You'll get the opposite issue that now it will pass all individual .o files to the shared lib instead of using .a archives, but that's probably fine.

To be honest, I like your change but I'm a bit scared because this thing is more subtle that it first appear. I'll try to test this PR with glib first, because it was stressing corner cases last time I worked on this.

Comment (xclaesse) [reply mesonbuild/interpreter/interpreter.py:3061]: Tried building GStreamer with everything as subproject, that's a huge stress test. It breaks in libnice, I get a bunch of
```
[93/155] Linking target nice/libnice.so.10.12.0
FAILED: nice/libnice.so.10.12.0 
cc  -o nice/libnice.so.10.12.0 agent/libagent.a.p/address.c.o agent/libagent.a.p/agent.c.o agent/libagent.a.p/candidate.c.o agent/libagent.a.p/component.c.o agent/libagent.a.p/conncheck.c.o agent/libagent.a.p/debug.c.o agent/libagent.a.p/discovery.c.o agent/libagent.a.p/inputstream.c.o agent/libagent.a.p/interfaces.c.o agent/libagent.a.p/iostream.c.o agent/libagent.a.p/outputstream.c.o agent/libagent.a.p/pseudotcp.c.o agent/libagent.a.p/stream.c.o agent/libagent.a.p/meson-generated_.._agent-enum-types.c.o socket/libsocket.a.p/socket.c.o socket/libsocket.a.p/udp-bsd.c.o socket/libsocket.a.p/tcp-bsd.c.o socket/libsocket.a.p/tcp-active.c.o socket/libsocket.a.p/tcp-passive.c.o socket/libsocket.a.p/pseudossl.c.o socket/libsocket.a.p/socks5.c.o socket/libsocket.a.p/http.c.o socket/libsocket.a.p/udp-turn.c.o socket/libsocket.a.p/udp-turn-over-tcp.c.o stun/libstun.a.p/stunagent.c.o stun/libstun.a.p/stunmessage.c.o stun/libstun.a.p/stun5389.c.o stun/libstun.a.p/stuncrc32.c.o stun/libstun.a.p/rand.c.o stun/libstun.a.p/stunhmac.c.o stun/libstun.a.p/utils.c.o stun/libstun.a.p/debug.c.o stun/libstun.a.p/usages_ice.c.o stun/libstun.a.p/usages_bind.c.o stun/libstun.a.p/usages_turn.c.o stun/libstun.a.p/usages_timer.c.o random/libnice-random.a.p/random.c.o random/libnice-random.a.p/random-glib.c.o -Wl,--as-needed -Wl,--no-undefined -shared -fPIC -Wl,--start-group -Wl,-soname,libnice.so.10 -Wl,--whole-archive agent/libagent.a socket/libsocket.a stun/libstun.a random/libnice-random.a -Wl,--no-whole-archive -Wl,--version-script,/home/xclaesse/programmation/gstreamer/subprojects/libnice/builddir/nice/libnice.map /usr/lib/x86_64-linux-gnu/libgio-2.0.so /usr/lib/x86_64-linux-gnu/libgobject-2.0.so /usr/lib/x86_64-linux-gnu/libglib-2.0.so /usr/lib/x86_64-linux-gnu/libgthread-2.0.so -pthread /usr/lib/x86_64-linux-gnu/libgnutls.so -Wl,--end-group
/usr/bin/ld: agent/libagent.a.p/meson-generated_.._agent-enum-types.c.o: in function `nice_component_state_get_type':
/home/xclaesse/programmation/gstreamer/subprojects/libnice/builddir/agent/agent-enum-types.c:13: multiple definition of `nice_component_state_get_type'; agent/libagent.a.p/meson-generated_.._agent-enum-types.c.o:/home/xclaesse/programmation/gstreamer/subprojects/libnice/builddir/agent/agent-enum-types.c:13: first defined here
...
```

I think this is because libnice does link_whole on a few internal static libraries: https://gitlab.freedesktop.org/libnice/libnice/-/blob/master/nice/meson.build#L22. As you can see in the command line above, you get for example libnice-random.a and all its .o files.

File: mesonbuild/interpreter/interpreter.py [hunk 2]
```diff
@@ -3039,39 +3039,34 @@ def add_target(self, name: str, tobj: build.Target) -> None:
...
-            # Keep only compilers used for linking
-            static_lib.compilers = {k: v for k, v in static_lib.compilers.items() if k in compilers.clink_langs}
+            # If we can reuse the object files from the static library, then we
+            # can simply `link_whole` with it, and use it's include_directories.
+            # We use the include_directoreis from the static_lib since it's
+            # already unrolled any include directories added by a dependency.
+            shared_lib = self.build_target(
+                node, [args[0]],
+                {'link_whole': [static_lib], 'include_directories': static_lib.get_include_dirs()},
+                build.SharedLibrary)
```
Comment (xclaesse) [inline mesonbuild/interpreter/interpreter.py:3067]: You need to pass most other kwargs too e.g. `install`, etc. `link_args` is tricky, because those could come from dependencies.

Why do you need include directories? This target should not be building any sources, so that should be useless, no?

**Comment by xclaesse** [comment]
LGTM

**Review by eli-schwartz — APPROVED** [review]

**Comment by dcbaker** [comment]
@jpakkane is this okay to go in during rc, or wait till 1.1?
